File: DSA/28Dec25/78.subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function (nums) {
    const ans = [[]];

    for (const num of nums) {
        const len = ans.length;
        for (let i = 0; i < len; i++) {
            // Clone the existing subset and add the new number
            ans.push([...ans[i], num]);
        }
    }

    return ans;
};

File: DSA/28Dec25/78.subsets.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        ans.push_back({}); // Start with empty subset
        
        for (int num : nums) {
            int n = ans.size();
            for (int i = 0; i < n; i++) {
                vector<int> newSubset = ans[i]; // Copy existing
                newSubset.push_back(num);       // Add current number
                ans.push_back(newSubset);       // Add to answer
            }
        }
        
        return ans;
    }
};
